specs/agents/tracing-spans-compress.md

```diff
@@ -0,0 +1,199 @@
+
+
+#### Configuration option `same_kind_compression_max_duration`
+
+Consecutive spans to the same destination that are under this threshold will be compressed into a single composite span.
+This reduces the collection, processing, and storage overhead, and removes clutter from the UI.
+The tradeoff is that the statement information will not be collected. 
+
+|                |          |
+|----------------|----------|
+| Type           | `duration`|
+| Default        | `1ms`    |
+| Dynamic        | `true`   |
+
+### Composite span
+
+Compressed spans don't have a physical span document.
+Instead, multiple compressed spans are represented by a composite span.
+
+#### Data model
+
+The `timestamp` and `duration` have slightly similar semantics,
+and they define properties under the `composite` context.
+
+- `timestamp`: The start timestamp of the first span.
+- `duration`: The sum of durations of all spans.
+- `composite`
+    - `count`: The number of compressed spans this composite span represents.
+      The minimum count is 2 as a composite span represents at least two spans.
+    - `end`: The end timestamp of the last compressed span.
+      The net duration of all compressed spans is equal to the composite spans' `duration`.
+      The gross duration (including "whitespace" between the spans) is equal to `compressed.end - timestamp`.
+
+#### Turning compressed spans into a composite span
+
+Spans have a `compress` method.
+The first time `compress` is called on a regular span, it becomes a composite span.
+
+```java
+void compress(Span other) {
+    if (compressed.count == 0) {
+        compressed.count = 2
+    } else {
+        compressedCount++
+    }
+    endTimestamp = max(endTimestamp, other.endTimestamp)
+}
+```
+
+#### Effects on metric processing
+
+As laid out in the [span destination spec](tracing-spans-destination.md#contextdestinationserviceresource),
+APM Server tracks span destination metrics.
+To avoid compressed spans to skew latency metrics and cause throughput metrics to be under-counted,
+APM Server will take `composite.count` into account when tracking span destination metrics.
+
+### Compression algorithm
+
+#### Eligibility for compression
+
+A span is eligible for compression if all the following conditions are met
+- It's an exit span
+- The trace context of this span has not been propagated to a downstream service
+
+The latter condition is important so that we don't remove (compress) a span that may be the parent of a downstream service.
+This would orphan the sub-graph started by the downstream service and cause it to not appear in the waterfall view.
+
+```java
+boolean isCompressionEligible() {
+    return exit && !context.hasPropagated
+}
+```
+
+#### Span buffering
+
+Non-compression-eligible spans may be reported immediately after they have ended.
+When a compression-eligible span ends, it does not immediately get reported.
+Instead, the span is buffered within its parent.
```

Comment:
we are buffering the raw spans, not the compressed, right?
So let's assume we applied same-kind-compression for a few consecutive spans, what's the name and duration of the currently buffered span? Is it the name and duration of the last raw span or the name and duration of the current composite span?

Reply:
In that case, the composite span would be buffered.